### PR TITLE
Support for numeric fluents and action costs

### PR DESCRIPTION
**Pull request by francescofuggitti**
## Proposed changes

In this PR, we add support for numeric fluents and action costs. It is built on top of the work originally proposed in #59. However, some updates on the lark domain and problem grammars and data structures have been applied to support all features of numeric fluents. 

This PR primarily addresses issues #38 and #89, but it also fixes the issues in #93 and #96.

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../blob/main/CONTRIBUTING.md) doc
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: tests/test_domain.py [hunk 1]
```diff
@@ -85,6 +96,113 @@ def test_build_simple_domain():
...
+    q = Predicate("q")
+    r = Predicate("r")
+    func1 = NumericFunction("f1", x, y)
+    func2 = NumericFunction("f2")
+    func3 = NumericFunction("f3")
+    action_1 = Action(
+        "action_1",
+        [x, y, z],
+        precondition=p
+        & FunctionEqualTo(func1, NumericValue(0))
```
Comment (haz) [inline tests/test_domain.py:129]: We should probably overload the expected operators to make the programmatic creation more succinct. Follow-up issue/PR?

**Review by haz — CHANGES_REQUESTED** [review]
Thank you so much for your efforts on this @francescofuggitti ! It's an epic addition. Especially by the diff count ;)

image

I suspect I'm missing some things that will come to light when we release this to the wild, but I've just made a few (mostly minor) comments/questions for you to consider.

File: pddl/_validation.py [hunk 2]
```diff
@@ -17,13 +17,31 @@
...
 from pddl.action import Action
+from pddl.custom_types import _check_not_a_keyword  # noqa: F401
 from pddl.custom_types import name as name_type
 from pddl.custom_types import namelike, to_names, to_types  # noqa: F401
 from pddl.exceptions import PDDLValidationError
 from pddl.helpers.base import check, ensure, ensure_set, find_cycle
 from pddl.logic import Predicate
 from pddl.logic.base import BinaryOp, QuantifiedCondition, UnaryOp
 from pddl.logic.effects import AndEffect, Forall, When
+from pddl.logic.functions import Assign, Decrease, Divide
```
[the remark this answers is not in the document]
Comment (francescofuggitti) [reply pddl/_validation.py:28]: I noticed that too, but apparently `isort` here doesn't like them to be on the tuple below. Don't know why.

File: pddl/logic/functions.py [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+        return self._terms
+
+    @property
+    def arity(self) -> int:
+        """Get the arity of the function."""
+        return len(self.terms)
+
+    # TODO: check whether it's a good idea...
+    # TODO: allow also for keyword-based replacement
+    # TODO: allow skip replacement with None arguments.
```
[the remark this answers is not in the document]
Comment (francescofuggitti) [reply pddl/logic/functions.py:57]: I got these from the previous commits, and I guess these are left over from the `predicates` module. Perhaps, we don't need them anymore

File: pddl/parser/common.lark [hunk 4]
```diff
@@ -37,6 +41,24 @@ IMPLY: "imply"
...
+METRIC: ":metric"
+ASSIGN: "assign"
+SCALE_UP: "scale-up"
+SCALE_DOWN: "scale-down"
+INCREASE: "increase"
+DECREASE: "decrease"
+MAXIMIZE: "maximize"
+MINIMIZE: "minimize"
+TOTAL_COST: "total-cost"
+NUMBER: /[0-9]+/
```
[the remark this answers is not in the document]
Comment (francescofuggitti) [reply pddl/parser/common.lark:61]: Good point, thanks! The data structure can deal with floats, but I just forgot to put them in here.

File: pddl/parser/domain.py [hunk 5]
```diff
@@ -23,10 +23,27 @@
 from pddl.helpers.base import assert_
 from pddl.logic.base import And, ExistsCondition, ForallCondition, Imply, Not, OneOf, Or
 from pddl.logic.effects import AndEffect, Forall, When
+from pddl.logic.functions import Assign, Decrease, Divide
```
[the remark this answers is not in the document]
Comment (francescofuggitti) [reply pddl/parser/domain.py:26]: will try again with `isort` and see what happens

File: pddl/parser/problem.lark [hunk 6]
```diff
@@ -21,6 +27,27 @@ goal:  LPAR GOAL gd_name RPAR
 gd_name:    atomic_formula_name
        |    LPAR NOT atomic_formula_name RPAR
        |    LPAR AND gd_name* RPAR
+       |    LPAR binary_comp f_exp f_exp RPAR
+
+f_exp:  NUMBER
```
[the remark this answers is not in the document]
Comment (francescofuggitti) [reply pddl/parser/problem.lark:32]: My first try was to use the rules from the domain, but didn't work as expected and just moved forward. Probably I was missing something...I'll iterate again on that.

File: pddl/_validation.py [hunk 7]
```diff
@@ -234,6 +252,86 @@ def _(self, predicate: Predicate) -> None:
...
+
+    @check_type.register
+    def _(self, times: Times) -> None:
+        """Check types annotations of a PDDL numeric times operator."""
+        self.check_type(times.operands)
+
+    @check_type.register
+    def _(self, divide: Divide) -> None:
+        """Check types annotations of a PDDL numeric divide operator."""
+        self.check_type(divide.operands)
```
Comment (marcofavorito) [inline pddl/_validation.py:333]: note, you could have used just `def _(self, binary_fun: BinaryFunction): ...`, the dispatch would have matched all subclasses of `BinaryFunction`

File: tests/test_domain.py (hunk 1 above)
Comment (francescofuggitti) [reply tests/test_domain.py:129]: Definitely! Given the amount of change here I guess it's better to do that in a follow-up PR

File: pddl/_validation.py (hunk 7 above)
Comment (francescofuggitti) [reply pddl/_validation.py:333]: You're right, thank you! I just forgot about the `BinaryFunction` and wrote all blindly 😅

Comment (marcofavorito) [reply pddl/_validation.py:333]: It's fine as you wrote it. In the future, we might add further validation for specific operators; hence, we might need (some of) those specific functions. Do what you wish :)

File: pddl/parser/domain.py [hunk 8]
```diff
@@ -98,6 +116,11 @@ def predicates(self, args):
         self._predicates_by_name = {p.name: p for p in predicates}
         return dict(predicates=predicates)
 
+    def functions(self, args):
+        """Process the 'functions' rule."""
+        function_definition = args[2]
```
[the remark this answers is not in the document]
Comment (francescofuggitti) [reply pddl/parser/domain.py:121]: Sure!

File: pddl/logic/functions.py [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+from pddl.parser.symbols import Symbols
+
+
+@cache_hash
+class FunctionExpression(Atomic):
+    """A class for all the function expressions."""
+
+
+@cache_hash
+@functools.total_ordering
```
Comment (marcofavorito) [inline pddl/logic/functions.py:31]: If I am not mistaken, we used `total_ordering` for any leaf expression in some expression hierarchy (in this case, function expression). `NumericFunction` is a leaf expression (correct?) so it is correctly annotated as with `total_ordering`.

What about the other _leaf_ expression below? `NumericValue`, `Metric` etc. If I recall correctly, tThe rationale is that: if instances of such types can appear in a sequence of some binary operator, then we should be able to sort them (if the operator is commutative). I guess this was useful in other places for canonicity.

File: pddl/_validation.py (hunk 7 above)
Comment (francescofuggitti) [reply pddl/_validation.py:333]: Yep, for now, I think it's better to wrap them all under the binary function so to have less code.

File: pddl/_validation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (francescofuggitti) [reply pddl/_validation.py:28]: Ok, got it. The alias for `EqualTo` is blocking `Assign, Decrease, Divide` from being added into the tuple. If we force `isort --combine-as` we get what we want.

File: pddl/logic/functions.py [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+    def __init__(self, *operands: FunctionExpression):
+        """
+        Initialize the Decrease operator.
+
+        :param operands: the operands.
+        """
+        super().__init__(*operands)
+
+
+class Minus(BinaryFunction, metaclass=MonotoneOp):
```
Comment (marcofavorito) [inline pddl/logic/functions.py:313]: I fear this is not correct. `MonotoneOp` will apply the idempotency simplification, which makes sense for logical `or` and `and`, but not for mathematical operations on numbers.

i.e. `a & a \equiv a`, but `a - a  \not\equiv a`.

You can see where this is implemented at this line: basically, we remove duplicates.
https://github.com/AI-Planning/pddl/blob/9b57554782f8748870aa4c255f451f0b0382a796/pddl/logic/base.py#L264




```suggestion
class Minus(BinaryFunction):
```

File: pddl/logic/functions.py [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+    def __init__(self, *operands: FunctionExpression):
+        """
+        Initialize the Minus operator.
+
+        :param operands: the operands.
+        """
+        super().__init__(*operands)
+
+
+class Plus(BinaryFunction, metaclass=MonotoneOp):
```
Comment (marcofavorito) [inline pddl/logic/functions.py:327]: see above on `Minus`

File: pddl/logic/functions.py [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+    def __init__(self, *operands: FunctionExpression):
+        """
+        Initialize the Plus operator.
+
+        :param operands: the operands.
+        """
+        super().__init__(*operands)
+
+
+class Times(BinaryFunction, metaclass=MonotoneOp):
```
Comment (marcofavorito) [inline pddl/logic/functions.py:341]: see above on `Minus`

File: pddl/logic/functions.py [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+    def __init__(self, *operands: FunctionExpression):
+        """
+        Initialize the Times operator.
+
+        :param operands: the operands.
+        """
+        super().__init__(*operands)
+
+
+class Divide(BinaryFunction, metaclass=MonotoneOp):
```
Comment (marcofavorito) [inline pddl/logic/functions.py:355]: see above on `Minus`

File: pddl/logic/functions.py (hunk 10 above)
Comment (francescofuggitti) [reply pddl/logic/functions.py:313]: That's right. I thought it was useful to flatten the operands.

Comment (marcofavorito) [reply pddl/logic/functions.py:313]: Minimum example to reproduce the problem:

```python
from pddl.logic.functions import Minus, NumericValue
a1 = NumericValue(1)
b1 = Minus(a1, a1, a1)  # this should be Minus, but we get NumericValue
a1 == b1 # we get True, but it should be false
```

Comment (marcofavorito) [reply pddl/logic/functions.py:313]: I think it should be as easy as removing the metaclass specification

**Review by marcofavorito — APPROVED** [review]
This is a monumental work @francescofuggitti, thank you so much. The tests witness that the feature is working (regardless of minor issues that might pop up in the future). Thanks!! :pray:

File: pddl/logic/functions.py (hunk 10 above)
Comment (francescofuggitti) [reply pddl/logic/functions.py:313]: great, I'll remove the metaclass!

File: pddl/logic/functions.py (hunk 9 above)
Comment (francescofuggitti) [reply pddl/logic/functions.py:31]: While `NumericValue` can appear in some binary operators, `Metric` does not. However, `Metric` is a leaf expression.

File: pddl/logic/functions.py (hunk 10 above)
Comment (marcofavorito) [reply pddl/logic/functions.py:313]: > That's right. I thought it was useful to flatten the operands.

this comment actually suggests that we should add a class attribute `idempotency: bool`, which by default is `False`, and change the behaviour of `MonotoneOp` depending on what is the value of `idempotency`. We should change the name for a generic binary operation metaclass, e.g. `BinaryOpMetaclass`. By doing so, we can still gain from the usefulness of flattening.

File: pddl/_validation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (francescofuggitti) [reply pddl/_validation.py:28]: Actually, if we combine the aliases in the tuple, then `flake8` starts complaining. I think, for the moment, it's better to leave them as it is.

File: pddl/logic/functions.py (hunk 10 above)
Comment (marcofavorito) [reply pddl/logic/functions.py:313]: #96

File: pddl/logic/base.py [hunk 14]
```diff
@@ -115,31 +115,34 @@ class Atomic(Formula):
...
 
     _absorbing: Optional[Formula] = None
+    idempotency: bool = False
 
     def __call__(cls, *args, **kwargs):
         """Init the subclass object."""
         operands = _simplify_monotone_op_operands(cls, *args)
-        if len(operands) == 1:
+        if len(operands) == 1 and cls.idempotency:
             return operands[0]
```
Comment (marcofavorito) [inline pddl/logic/base.py:128]: great, but I expected also line 126, the call to  `_simplify_monotone_op_operands`, to change.

Now this gives me:
```python
from pddl.logic.functions import Minus, NumericValue
a1 = NumericValue(1)
m = Minus(a1, a1, a1)
m.operands  # returns a tuple of one element: `(a1, )`
```

File: pddl/logic/base.py [hunk 15]
```diff
@@ -115,31 +115,34 @@ class Atomic(Formula):
...
-class MonotoneOp(type):
+class BinaryOpMetaclass(type):
     """Metaclass to simplify monotone operator instantiations."""
 
     _absorbing: Optional[Formula] = None
+    idempotency: bool = False
 
     def __call__(cls, *args, **kwargs):
         """Init the subclass object."""
         operands = _simplify_monotone_op_operands(cls, *args)
```
Comment (marcofavorito) [inline pddl/logic/base.py:126]:
```suggestion
        operands = _simplify_monotone_op_operands(cls, *args, idempotency=cls.idempotency)
```

Comment (marcofavorito) [reply pddl/logic/base.py:126]: Then:


```python
def _simplify_monotone_op_operands(cls, *operands, idempotency: bool = False):
    if idempotency:
        operands = list(dict.fromkeys(operands))
    ...
```

File: pddl/logic/base.py (hunk 14 above)
Comment (francescofuggitti) [reply pddl/logic/base.py]: right! This has been fixed in [2682e2f](https://github.com/AI-Planning/pddl/pull/92/commits/2682e2f51f89b920f697896ae485e58b6d4f3296)

File: pddl/parser/problem.lark (hunk 6 above)
[the remark this answers is not in the document]
Comment (francescofuggitti) [reply pddl/parser/problem.lark]: Addressed!

File: pddl/parser/problem.py [hunk 16]
```diff
@@ -130,6 +190,65 @@ def atomic_formula_name(self, args):
             ]
             return Predicate(name, *terms)
 
+    def f_exp(self, args):
```
[the remark this answers is not in the document]
Comment (francescofuggitti) [reply pddl/parser/problem.py:193]: Done!

File: pddl/logic/functions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (haz) [reply pddl/logic/functions.py]: Ya, I'd suggest nixing them. Hidden `TODO`'s are just where future ideas go to fade away ;)

File: pddl/_validation.py (hunk 7 above)
Comment (haz) [reply pddl/_validation.py]: This resolved?

**Review by haz — APPROVED** [review]
Just some open question on what's all resolved. Rest for @marcofavorito to sign off on.

**Review by marcofavorito — APPROVED** [review]
LGTM

**Pull request merged**